Specialise `any` and `all` for tuples

jipolanco opened the pull request:
This avoids a small allocation (and improves performance) when iterating over a `ProductIterator` constructed using mixed iterator types. See #40283 for details.

I couldn't really identify where the allocation comes from, and there's likely an underlying issue to be still fixed. But this does solve the `ProductIterator` issue.

When used with homogeneous iterator types, this has no additional cost compared to the original short-circuiting version, as seen in the example below.

Also note that `Base.map` (used in this PR) was already being used in `isdone(z::Zip)`.

Closes #40283.

Example:

```julia
using BenchmarkTools

# 1. Mixing different iterator types (heterogeneous iterator)
xs = (1:5, 1:0.5:4)
iter = Iterators.product(xs...)

# Julia master:
@btime iterate($iter);  # 37.649 ns (2 allocations: 144 bytes)

# With this PR:
@btime iterate($iter);  # 8.586 ns (0 allocations: 0 bytes)

# 2. Using the same iterator type (homogeneous iterator)
xs = (1:5, 2:4)
iter = Iterators.product(xs...)

# Julia master:
@btime iterate($iter);  # 2.181 ns (0 allocations: 0 bytes)

# With this PR:
@btime iterate($iter);  # 2.124 ns (0 allocations: 0 bytes)
```

N5N3 commented:
`any(f, iter::Tuple)` use loop-based fallback thus any mixed types would do harm to performance.
Maybe we should force inline `iter` with small length (similar to `Base._tuple_any`)

simeonschaub commented on base/iterators.jl:
```diff
@@ -1032,7 +1032,7 @@ _prod_eltype(::Type{I}) where {I<:Tuple} =
 iterate(::ProductIterator{Tuple{}}, state) = nothing
 
-@inline isdone(P::ProductIterator) = any(isdone, P.iterators)
+@inline isdone(P::ProductIterator) = any(Base.map(isdone, P.iterators))  # issue #40283
```
We should make sure to keep the sort-circuiting behavior here and not unconditionally evaluate `isdone` for every entry in `P.iterators`. I agree with @N5N3 that there should probably be an optimized version of `any` for this.

jipolanco replied:
Got it, in that case we can add a version of `any` for tuples. I guess the same should be done for `all`, which seems to have the same issue.

jipolanco commented:
Seems reasonable to add an optimised version of `any` (and also `all`?) for tuples.

Here is one possible implementation:

```julia
# any(f, itr) = _any(f, itr, :)  # definition in base/reduce.jl

_any(f, itr::Tuple, ::Colon) = _any_tuple(f, false, itr...)

@inline function _any_tuple(f, anymissing, x, rest...)
    v = f(x)
    if ismissing(v)
        anymissing = true
    elseif v
        return true
    end
    return _any_tuple(f, anymissing, rest...)
end

@inline _any_tuple(f, anymissing) = anymissing ? missing : false
```

Should this be limited to small tuples?

jipolanco committed:
Specialise `any` and `all` for tuples

jipolanco committed:
Add `any` and `all` tests using `missing`

jipolanco committed:
Add allocation tests

N5N3 commented:
I think it's OK to put these optimization in `reduce.jl` to fix the build error.
Another suggestion: if `eltype(x::Tuple)` is concrete, it seems good to use for loop and Let LLVM unroll for us.
```julia
# If eltype(x) is concrete. Use for loop
any(f, x::NTuple{N}) where {N} = _any(f, x, :)
# Otherwise force unroll to avoid union-split. (Only for small Tuple)
function any(f, x::NTuple{N,Any}) where {N}
    N >= 32 && return _any(f, x, :) # Not tuned, just follow Any32
    _any_tuple(f, false, x...)
end
```

jipolanco committed:
Move specialisations to reduce.jl

jipolanco committed:
Fall back to loop implementation in certain cases

Fall back to original for-loop implementation if the tuple either:

- is a homogeneous tuple (all elements have the same type);
- has length > 32.

jipolanco commented:
@NSN3 Thanks for the suggestions! This should be taken care of now.

KristofferC committed:
Merge branch 'master' into jip/product-iterator

KristofferC commented:
@jipolanco Is this good to go?
